File: 2022/day_06.cpp
//
// Created by Wollenschneider Luis on 04.12.22.
//

#include "../utils.hpp"
#include <list>

#define DAY "06"
#define YEAR "2022"

std::string parse_input(const std::string& filepath) {
    // check if path exists
    std::string input;
    std::ifstream file(filepath);
    if (!file.is_open()) {
        std::cout << "File not found" << std::endl;
        return input;
    }
    std::string line;
    while (std::getline(file, line)) {
        input = line;
    }
    file.close();
    return input;
}

int distinct(const std::string& inp, int n) {
    char* last_n = new char[n];
    int i;
    for (i = 0; i < n-1; i++) {
        last_n[i] = inp[i];
    }
    for (; i < inp.length(); i++) {
        last_n[n-1] = inp[i];
        bool found = true;
        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                if (j != k && last_n[j] == last_n[k]) {
                    found = false;
                    break;
                }
            }
            if (!found) {
                break;
            }
        }
        if (found) {
            return i + 1;
        }
        for (int j = 0; j < n-1; j++) {
            last_n[j] = last_n[j+1];
        }
    }
    return -1;
}

int part1(const std::string& inp) {
    return distinct(inp, 4);
}

int part2(const std::string& inp) {
    return distinct(inp, 14);
}

int main() {
    std::string inp = parse_input(YEAR "/inputs/day_" DAY ".txt");
    std::string test_inp = parse_input(YEAR "/tests/day_" DAY ".txt");

    int expected_result_part1 = get_expected_result<int>(YEAR "/tests/results/day_" DAY "_1.txt");

    int res = part1(test_inp);
    bool eval_res = evaluate_results(res, expected_result_part1);
    res = part1(inp);
    std::cout << "Part 1: " << res << std::endl;

    if (!eval_res) {
        return 1;
    }

    std::cout << std::endl;
    int expected_result_part2 = get_expected_result<int>(YEAR "/tests/results/day_" DAY "_2.txt");

    res = part2(test_inp);
    evaluate_results(res, expected_result_part2);
    res = part2(inp);
    std::cout << "Part 2: " << res << std::endl;

    return 0;
}

File: 2025/day_06.ts
import { solve } from '../utils';

const YEAR = 2025;
const DAY = 6;

function parseInput(input: string): any {
    return input.split('\n').map(line => line.trim().split(' ').filter(c => c.length > 0));
}


function part1(input: string): any | null {
    let inp = parseInput(input);
    let s = 0;
    for (let i = 0; i < inp[0].length; i++) {
        let ns = [];
        for (let j = 0; j < inp.length - 1; j++) {
            ns.push(Number(inp[j][i]));
        }
        if (inp[inp.length - 1][i] === '*') {
            s += ns.reduce((a, b) => a * b, 1);
        } else if (inp[inp.length - 1][i] === '+') {
            s += ns.reduce((a, b) => a + b, 0);
        }
    }
    return s;
}

function part2(input: string): any | null {
    let s = 0;
    let lines = input.split('\n');
    for (let j = 0; j < lines.length; j++) {
        lines[j] += ' ';
    }
    let ns = [];
    let op = null;
    for (let i = 0; i < lines[0].length; i++) {
        let n = '';
        for (let j = 0; j < lines.length - 1; j++) {
            n += lines[j][i];
        }
        if (op === null) {
            op = lines[lines.length - 1][i];
        }
        if (n.trim().length > 0) {
            ns.push(Number(n));
        } else {
            if (op === '*') {
                s += ns.reduce((a, b) => a * b, 1);
            } else if (op === '+') {
                s += ns.reduce((a, b) => a + b, 0);
            }
            ns = [];
            op = null;
        }
    }

    return s;
}

await solve(part1, YEAR, DAY, 1);
await solve(part2, YEAR, DAY, 2);